PEP 775: Make zlib required to build CPython (#4281)
python/peps: peps/pep-0775.rst

PEP: 775
Title: Make zlib required to build Python
Author: Stan Ulbrych <stanulbrych@gmail.com>,
        Petr Viktorin <encukou@gmail.com>
Discussions-To: https://discuss.python.org/t/lets-make-zlib-required-rather-than-optional-to-build-cpython/23062
Status: Draft
Type: Standards Track
Created: 24-Feb-2025
Python-Version: 3.14
Post-History: `23-Jan-2023 <https://discuss.python.org/t/23062>`__

Abstract
========

Building CPython on systems without the `zlib <https://zlib.net>`_ compression library, except WASI,
is no longer supported.

The :mod:`zlib` module is made a required part of the standard library,
except on WASI.

Builds without zlib will still be possible, but formally unsupported.

Motivation
==========

The zlib library, which powers the :mod:`!zlib` Python module,

> is available on virtually all systems.

Author: There is no source that would list them all that I know of, and i agree "virtually all" is subjective. The [zlib FAQ ](https://www.zlib.net/zlib_faq.html) seems to cover some. What do you think Petr?

> is available on virtually all systems.

Reviewer: Don't know of a relevant source that would back that up. I also don't know a system that can run CPython but doesn't have zlib readily available.

Perhaps add a footnote that says something like

- `zlib` is available on all supported systems except WASI
- `zlib` is a dependency of Git
- `zlib` is the [de-facto standard](https://en.wikipedia.org/wiki/Zlib#Applications) library for the DEFLATE algorithm, required for things like opening compressed `zip` files or decoding PNG images

> is available on virtually all systems.

Reviewer 2: zlib (used for PNGs) has been required by default for [building Pillow](https://pillow.readthedocs.io/en/stable/installation/building-from-source.html) for [~10 years](https://pillow.readthedocs.io/en/stable/releasenotes/3.0.0.html).

> is available on virtually all systems.

Reviewer: Not just building. Like many* projects on PyPI, you need zlib to unpack Pillow's wheels and sdists.

(Also, before building you'd need to get the source. It's *theoretically* possible, but not easy, to avoid `zlib` in Git, compressed archives, or HTTP transfer compression).

(\* here too I'd say *virually all*, but I don't have a published source)

.. _pip: https://pypi.org/project/pip/

CPython developers don't really notice either. It turns out that at the time
of this writing, at least one CPython test fails without ``zlib`` (the “skip”
decorator in ``test_peg_generator.test_c_parser`` is applied too late),
but our CI didn't catch this.

This PEP treats this as an issue in documentation and messaging.
In practice, we already don't support building CPython without ``zlib``; we
should just say so.

Rationale
=========

There are possible use cases for zlib-less builds, such as embedding and
bootstrapping, as well as unforeseen ones.
Therefore, we don't *remove* support for zlib-less systems; we mark them
unsupported and invite affected users to do their own testing, or to share
use cases that can make us reconsider this decision.

``zlib`` is not yet used by default on the WASI platform -- mostly because
adding it hasn't yet been a priority there. (Note that `Pyodide`_, the main
“real-world” CPython distribution for WASI, does include ``zlib``.)
We take this as an opportunity to  continue testing a platform without
``zlib``, so that we don't unintentionally break unsupported builds yet.

> 

Reviewer: ```suggestion

.. _Pyodide: https://pyodide.org

```

Specification
=============

In standard library modules that use zlib for optional functionality,
that functionality will raise ``ImportError`` when used.

> Code to generate more “friendly” error messages, or to pre-check whether

Reviewer 3: ```suggestion
Code to generate more "friendly" error messages, or to pre-check whether
```

zlib is available, will be removed.
All functionality unrelated to zlib will still be usable if zlib is
missing.

This affects the following modules, and more that depend on these
transitively:

* :mod:`shutil` (``gztar`` and ``zip`` archive formats)

> * :mod:`tarfile`, :mod:`zipfile`, :mod:`zipimport`, :mod:`zipapp` (archive compression)

Reviewer 3: ```suggestion
* :mod:`tarfile`, :mod:`zipfile`, :mod:`zipimport`,
  :mod:`zipapp` (archive compression)
```

* :mod:`codecs` (``zlib_codec``)

``shutil.get_archive_formats()`` will always include ``zip`` and ``gztar``
as registered formats, even if they are unusable due to missing zlib.

The ``configure`` script will issue a warning when zlib is not found on
platforms other than WASI.

``test_zlib`` will fail on platforms other than WASI.
All other tests will continue to be skipped -- that is, uses of
``@test.support.requires_zlib`` will be kept in place -- for the benefit
of WASI, unsupported builds, and any possible reverts.

:pep:`11` will be adjusted to mark "Systems without zlib, except WASI" as
unsupported.

Backwards Compatibility
=======================

In practice, nothing major changes, except in error cases -- for example,
attempts to use tar compression without zlib available will raise
``ImportError`` and not ``CompressionError``.

Security Implications
=====================

None known.

How to Teach This
=================

We don't expect that any instructions will need to change, as zlib is
already available in all relevant contexts.

Reference Implementation
========================

> https://github.com/python/cpython/pull/130297

Reviewer 3: ```suggestion
A reference implementation may be found in a pull request to the CPython
repository, `python/cpython#130297
<https://github.com/python/cpython/pull/130297>`_
```

PEP: 775
Title: Make zlib required to build Python
Author: Stan Ulbrych <stanulbrych@gmail.com>,
        Petr Viktorin <encukou@gmail.com>
Discussions-To: https://discuss.python.org/t/lets-make-zlib-required-rather-than-optional-to-build-cpython/23062
Status: Draft

> Type: Standards Track

Reviewer 2: This PEP mentions CPython in the abstract, motivation and rationale.

The specification doesn't mention CPython or any other implementations such as PyPy, IronPython, MicroPython and Jython.

Does this PEP only apply to CPython? Should other implementations be acknowledged, even if only to explicitly say this only applies to CPython? If so, any advice to them?

Author: This is for building cpython. Requirements for building other implementations are up to their maintainers, there are already differences between cpython and implementations when it comes to zlib for example:

<details>
<summary>PyPy already lists zlib as a dependency.</summary>

![image](https://github.com/user-attachments/assets/2d8ac391-c5f2-4052-8a65-1365e662708b)

</details>

The Motivation section explains why and I would say is good enough advice.

> Created: 20-Feb-2025

Reviewer 2: This is the date the PEP number was assigned:
```suggestion
Created: 24-Feb-2025
```

Python-Version: 3.14
Post-History: `23-Jan-2023 <https://discuss.python.org/t/23062>`__

Abstract
========

Building CPython on systems without the `zlib <https://zlib.net>`_ compression library, except WASI,
is no longer supported.

> The ``zlib`` module is made a required part of the standard library,

Reviewer 2: ```suggestion
The :mod:`zlib` module is made a required part of the standard library,
```

except on WASI.

> Builds without ``zlib`` will still be possible, but formally unsupported.

Reviewer 2: ```suggestion
Builds without zlib will still be possible, but formally unsupported.
```

Motivation
==========

> The ``zlib`` library, which powers the ``zlib`` Python module,

Reviewer 2: The zlib compression library doesn't need code formatting (re: https://zlib.net, https://en.wikipedia.org/wiki/Zlib)

```suggestion
The zlib library, which powers the :mod:`!zlib` Python module,
```

is available on virtually all systems.

> Many wheels on PyPI, including the `pip`_ installer, require ``zlib``.
> Users of *pip* would consider CPython without ``zlib`` to be broken,
> but mostly don't notice because all major builds of CPython include ``zlib``.
> 
> .. _pip: https://pypi.org/project/pip/

Reviewer 2: ```suggestion
Many wheels on PyPI, including the :pypi:`pip` installer, require zlib.
Users of pip would consider CPython without zlib to be broken,
but mostly don't notice because all major builds of CPython include zlib.
```

> CPython developers don't really notice either. It turns out that at the time
> of this writing, at least one CPython test fails without ``zlib`` (the “skip”
> decorator in ``test_peg_generator.test_c_parser`` is applied too late),
> but our CI didn't catch this.

Reviewer 2: ```suggestion
CPython developers don't really notice either. It turns out that at the time
of writing, at least one CPython test fails without zlib (the "skip"
decorator in ``test_peg_generator.test_c_parser`` is applied too late),
but our CI didn't catch this.
```

This PEP treats this as an issue in documentation and messaging.

> In practice, we already don't support building CPython without ``zlib``; we

Reviewer 2: ```suggestion
In practice, we already don't support building CPython without zlib; we
```

should just say so.

Rationale
=========

There are possible use cases for zlib-less builds, such as embedding and
bootstrapping, as well as unforeseen ones.
Therefore, we don't *remove* support for zlib-less systems; we mark them
unsupported and invite affected users to do their own testing, or to share
use cases that can make us reconsider this decision.

> ``zlib`` is not yet used by default on the WASI platform -- mostly because
> adding it hasn't yet been a priority there. (Note that `Pyodide`_, the main
> “real-world” CPython distribution for WASI, does include ``zlib``.)
> We take this as an opportunity to  continue testing a platform without
> ``zlib``, so that we don't unintentionally break unsupported builds yet.

Reviewer 2: ```suggestion
zlib is not yet used by default on the WASI platform -- mostly because
adding it hasn't yet been a priority there. (Note that `Pyodide`_, the main
"real-world" CPython distribution for WASI, does include zlib.)
We take this as an opportunity to  continue testing a platform without
zlib, so that we don't unintentionally break unsupported builds yet.
```

.. _Pyodide: https://pyodide.org

Specification
=============

> In standard library modules that use ``zlib`` for optional functionality,
> that functionality will raise ``ImportError`` when used.
> Code to generate more “friendly” error messages, or to pre-check whether
> ``zlib`` is available, will be removed.
> All non-``zlib``-related functionality will still be usable if ``zlib`` is
> missing.

Reviewer 2: ```suggestion
In standard library modules that use zlib for optional functionality,
that functionality will raise ``ImportError`` when used.
Code to generate more “friendly” error messages, or to pre-check whether
zlib is available, will be removed.
All functionality unrelated to zlib will still be usable if zlib is
missing.
```

This affects the following modules, and more that depend on these
transitively:

> * ``shutil`` (``gztar`` and ``zip`` archive formats)
> * ``tarfile``, ``zipfile``, ``zipimport``, ``zipapp`` (archive compression)
> * ``codecs`` (``zlib_codec``)

Reviewer 2: Optionally, if you want to link to the modules:

```suggestion
* :mod:`shutil` (``gztar`` and ``zip`` archive formats)
* :mod:`tarfile`, :mod:`zipfile`, :mod:`zipimport`, :mod:`zipapp` (archive compression)
* :mod:`codecs` (``zlib_codec``)
```

``shutil.get_archive_formats()`` will always include ``zip`` and ``gztar``

> as registered formats, even if they are unusable due to missing ``zlib``.
> 
> The ``configure`` script will issue a warning when ``zlib`` is not found on
> platforms other than WASI.

Reviewer 2: ```suggestion
as registered formats, even if they are unusable due to missing zlib.

The ``configure`` script will issue a warning when zlib is not found on
platforms other than WASI.
```

``test_zlib`` will fail on platforms other than WASI.
All other tests will continue to be skipped -- that is, uses of
``@test.support.requires_zlib`` will be kept in place -- for the benefit
of WASI, unsupported builds, and any possible reverts.

> PEP 11 will be adjusted to mark "Systems without zlib, except WASI" as

Reviewer 2: ```suggestion
:pep:`11` will be adjusted to mark "Systems without zlib, except WASI" as
```

unsupported.

Backwards Compatibility
=======================

In practice, nothing major changes, except in error cases -- for example,

> attempts to use tar compression withoutt ``zlib`` available will raise

Reviewer 2: ```suggestion
attempts to use tar compression without zlib available will raise
```

``ImportError`` and not ``CompressionError``.

Security Implications
=====================

None known.

How to Teach This
=================

> We don't expect that any instructions will need to change, as ``zlib`` is

Reviewer 2: ```suggestion
We don't expect that any instructions will need to change, as zlib is
```

already available in all relevant contexts.

Reference Implementation
========================

https://github.com/python/cpython/pull/130297

Future work
===========

> In the future, if no use cases for ``zlib``-less  builds are found,
> ``zlib`` may be made fully required.

Reviewer 2: ```suggestion
In the future, if no use cases for zlib-less builds are found,
zlib may be made fully required.
```

PEP: 775

> Title: Make zlib required to build Python

Reviewer 3: Perhaps?
```suggestion
Title: Make zlib required to build CPython
```

> Author: Gregory P. Smith <greg@krypto.org>,

Reviewer 3: cc @gpshead for confirmation
